### ISSUE TITLE
Cask/HomepageMatchesUrl check is too trigger happy

### ISSUE THREAD

**Issue by lenlo** [Homebrew/brew#5202]
**Please note we will close your issue without comment if you delete, do not read or do not fill out the issue checklist below and provide ALL the requested information. If you repeatedly fail to use the issue template, we will block you from ever submitting issues to Homebrew again.**

- [x] are reporting a bug others will be able to reproduce and not asking a question. If you're not sure or want to ask a question do so on our Discourse: https://discourse.brew.sh
- [ ] ran a `brew` command and reproduced the problem with multiple formulae? If it's a problem with a single, official formula (not cask) please file this issue at Homebrew/homebrew-core: https://github.com/Homebrew/homebrew-core/issues/new/choose. If it's a `brew cask` problem please file this issue at https://github.com/Homebrew/homebrew-cask/issues/new/choose. If it's a tap (e.g. Homebrew/homebrew-php) problem please file this issue at the tap.
- [ ] ran `brew update` and can still reproduce the problem?
- [ ] ran `brew doctor`, fixed all issues and can still reproduce the problem?
- [ ] ran `brew config` and `brew doctor` and included their output with your issue?

To help us debug your issue please explain:
- What you were trying to do (and why)
- What happened (include command output)
- What you expected to happen
- Step-by-step reproduction instructions (by running `brew` commands)

I was running `cask-repair` on a formula and among other things added a comment just before the URL line, but when I exited the editor, I got an error saying that the comment was unnecessary and the change therefore was rejected:

```
diff --git a/Casks/vuescan.rb b/Casks/vuescan.rb
index d440a39961..f9b169153f 100644
--- a/Casks/vuescan.rb
+++ b/Casks/vuescan.rb
@@ -2,7 +2,9 @@ cask 'vuescan' do
   version '9.6.20'
   sha256 :no_check # required as upstream package is updated in-place
 
-  url "https://www.hamrick.com/files/vuex64#{version.major_minor.no_dots}.dmg"
+  # Include a dummy query in the URL to force homebrew to download a new copy of the
+  # file when the patch level changes and not just reuse any previously cached copy.
+  url "https://www.hamrick.com/files/vuex64#{version.major_minor.no_dots}.dmg?#{version.patch}"
   appcast 'https://www.hamrick.com/alternate-versions.html'
   name 'VueScan'
   homepage 'https://www.hamrick.com/'
----------------------------------------------------------------------------------------------------------
• == /opt/homebrew/Library/Taps/homebrew/homebrew-cask/Casks/vuescan.rb ==
C:  6:  3: Cask/HomepageMatchesUrl: hamrick.com matches hamrick.com, the comment above the url stanza is unnecessary

1 file inspected, 1 offense detected
----------------------------------------------------------------------------------------------------------
```

The thing is -- as you can see -- `hamrick.com` wasn't in the comment. I scratched my head for a while until I was able to dig up the source where this check was being performed. `rubocop-cask/lib/rubocop/cop/cask/homepage_matches_url.rb` has:

```
        def add_offense_unnecessary_comment(stanza)
          return unless comment?(stanza)
          return unless url_match_homepage?(stanza)
          return unless comment_matches_url?(stanza)

          comment = comment(stanza).loc.expression
          add_offense(comment,
                      location: comment,
                      message: format(MSG_UNNECESSARY, domain: domain(stanza), homepage: homepage))
        end

        def url_from_comment(stanza)
          comment(stanza).text
            .sub(/[^ ]*# ([^ ]+) .*/, '\1')
        end

        def comment_matches_url?(stanza)
          full_url(stanza).include?(url_from_comment(stanza))
        end
```

In other words, the comment will be flagged if the first word after the # can be found anywhere in the URL. In my case `file` can be found in `https://www.hamrick.com/files/vuex64#{version.major_minor.no_dots}.dmg?#{version.patch}`, so the formula was rejected. This is clearly too simplistic.

I'm guessing that this check has some kind of historic significance because it seems unlikely that a lot of people would include the URL in the comment before the URL. Is it still needed? Otherwise, I vote to remove this code instead of trying to make it more accurate and possibly failing at that. (@reitermarkus?)

**Comment by commitay** [comment]
> Is it still needed?

Yes, it's required in > 2300 casks.

**Comment by lenlo** [comment]
> > Is it still needed?
>
> Yes, it's required in > 2300 casks.

In what way? I would love to understand this better. Why is it important to make sure that the comment does not mention the URL? Is it used for some functional purpose beside just being a comment? Or am I misunderstanding the purpose of this check?

**Comment by commitay** [comment]
The check ensures that only casks that require this comment have one.

https://github.com/Homebrew/homebrew-cask/blob/4c6856184f8cd151f64488deb5f7ffbf8f449f77/Casks/atom.rb#L5-L6
```
  # github.com/atom/atom was verified as official when first introduced to the cask
  url "https://github.com/atom/atom/releases/download/v#{version}/atom-mac.zip"
```

**Comment by commitay** [comment]
IIRC, this is the first time that it has flagged a comment that would otherwise be considered acceptable.

I don't see the need to make any changes here as comments apart from the "was verified as official" are very unusual in casks.

**Comment by lenlo** [comment]
> The check ensures that only casks that require this comment have one.
> 
> https://github.com/Homebrew/homebrew-cask/blob/4c6856184f8cd151f64488deb5f7ffbf8f449f77/Casks/atom.rb#L5-L6
> 
> ```
>   # github.com/atom/atom was verified as official when first introduced to the cask
>   url "https://github.com/atom/atom/releases/download/v#{version}/atom-mac.zip"
> ```

So you want cask formulas that have different host names in the `url` and `homepage` to have a comment documenting this as per [Cask Language Reference/Stanzas/URL](https://github.com/Homebrew/homebrew-cask/blob/master/doc/cask_language_reference/stanzas/url.md). I get that. But what's the point of rejecting other formulas that happen to mention parts of the `url` in the comment too? That doesn't make any sense to me. It might be for any reason and not related to this documenting requirement at all.

> IIRC, this is the first time that it has flagged a comment that would otherwise be considered acceptable.
> 
> I don't see the need to make any changes here as comments apart from the "was verified as official" are very unusual in casks.

I beg to differ. If (*if*) it's important to make sure that a "was verified as official" comment isn't included unnecessarily, then let's check for exactly those words (assuming that the wording really is fixed). Let's not just randomly reject other comments with misleading error messages that confuses people and causes frustration. The check is clearly too broad today; let's improve it and make it better.

**Comment by lenlo** [comment]
Sorry, I just realized that I really should have filed this under `Homebrew/rubocop-cask` instead of `Homebrew/brew`. My apologies. Let's see if I can get it moved...
